File: getRows.m
% Finds all rows with more than one domino and their sums
% 
% Rows = cell[numRows x 1]
%
% row = [2 x dominosInRow]
%    - row[1, i] = domino number
%    - row[2, i] = domino orientation w/ respect to row
%           1 = horizontal
%           2 = Top Vertical
%           3 = Bottom Vertical
%           4 = Middle Vertical
%
% 

function [rows, sums, rowEnds] = getRows(dominos, adjacentDoms)

[numDominos, ~] = size(dominos);

rows = cell(numDominos, 1);
numRows = 0;

% Get Rows

for i = 1 : numDominos
    currDom = i;
    thisDom = dominos(i, :);
    left = i;
    orient = -1;
    nextOrient = -1;
    
    % Move left until we reach the beginning of the column.
    
    while (left ~= 0)
        currDom = left;
        orient = nextOrient;
        thisDom = dominos(left, :);
        if (thisDom(5) == 1) 
            [left, nextOrient] = leftDom(dominos, adjacentDoms, currDom, orient);
            
            if (left > 0)
                currDom = left;
            end
        else 
            if (orient ~= -1)
                [left, nextOrient] = leftDom(dominos, adjacentDoms, currDom, orient);
            else
                [left, nextOrient] = leftDom(dominos, adjacentDoms, currDom, 2);
                
                if (left == 0)
                    [left, nextOrient] = leftDom(dominos, adjacentDoms, currDom, 3);
                end
                if (left == 0)
                    [left, nextOrient] = leftDom(dominos, adjacentDoms, currDom, 4);
                end
            end
        end
    end
             
    % Move right until we reach the end of the row
    % If there is only one domino in the row, don't add it to the row
    % matrix. After adding the next domino to the row, delete the
    % connection from the adjacentDoms matrix.
    
    thisRow = zeros(2,numDominos);
    dominosInRow = 1;
    
    
    if (orient ~= -1)
        thisRow(1,1) = currDom;
        thisRow(2,1) = orient;
        [right, nextOrient] = rightDom(dominos, adjacentDoms, currDom, orient);
        
    else
        thisRow(1,1) = currDom;
        
        [right, nextOrient] = rightDom(dominos, adjacentDoms, currDom, 1);
        thisRow(2,1) = 1;
        
        if (right == 0)
            
            [right, nextOrient] = rightDom(dominos, adjacentDoms, currDom, 2);
             thisRow(2,1) = 2;
        end
        
        if (right == 0)
            [right, nextOrient] = rightDom(dominos, adjacentDoms, currDom, 3);
            thisRow(2,1) = 3;
        end
        if (right == 0)
            [right, nextOrient] = rightDom(dominos, adjacentDoms, currDom, 4);
            thisRow(2,1) = 4;
        end
    end
    
    % There is a multi-domino row
    if (right > 0)
        
        while(right > 0)
            % Set adjacencies of dominos already in the row to zero.
            adjacentDoms(currDom, right) = 0;
            adjacentDoms(right, currDom) = 0;
            
            currDom = right;
            orient = nextOrient;
            dominosInRow = dominosInRow + 1;
            
            thisRow(1, dominosInRow) = currDom;
            thisRow(2, dominosInRow) = orient;
            
            [right, nextOrient] = rightDom(dominos, adjacentDoms, currDom, orient);
            
        end
        
        newRow = zeros(2, dominosInRow);
        
        for j = 1 : dominosInRow
            newRow(1,j) = thisRow(1,j);
            newRow(2,j) = thisRow(2,j);
        end
        
        numRows = numRows + 1;
        rows{numRows} = newRow;
        
    end  
end

temp = cell(numRows, 1);
for i = 1 : numRows
    temp{i} = rows{i};
end

rows = temp;

sums = zeros(numRows, 1);
rowEnds = zeros(numRows, 2); % [first, last]

% Find sums and end values of each row
for i = 1 : numRows
    currRow = rows{i};
    [~, numDoms] = size(currRow);
    
    sum = 0;
    
    for j = 1 : numDoms
        currDom = dominos(currRow(1,j), :);
        currOrient = currRow(2,j);
        
        if (currOrient == 1)
            rowVal = currDom(6) + currDom(7);
            startVal = currDom(6);
            endVal = currDom(7);
        end
        if (currOrient == 2)
            rowVal = currDom(6);
            startVal = currDom(6);
            endVal = currDom(6);
        end
        if (currOrient == 3)
            rowVal = currDom(7);
            startVal = currDom(7);
            endVal = currDom(7);
        end
        if (currOrient == 4)
            rowVal = currDom(6) + currDom(7);
            startVal = currDom(6);
            endVal = currDom(6);
        end
        
        sum = sum + rowVal;
        
        if (j == 1)
            rowEnds(i, 1) = startVal;
        end
        if (j == numDoms)
            rowEnds(i, 2) = endVal;
        end
        
    end
    
    sums(i) = sum;
    
end

end

File: python/domino_helper/get_rows.py
"""
getRows.m -> get_rows.py

Finds all rows of connected dominos and computes their sums.

Row orientation codes:
    1 = horizontal
    2 = Top Vertical
    3 = Bottom Vertical
    4 = Middle Vertical
"""

import numpy as np
from .left_dom import left_dom
from .right_dom import right_dom


def get_rows(dominos, adjacent_doms):
    """
    Find all rows with more than one domino and compute their sums.

    Parameters
    ----------
    dominos : np.ndarray
        (num_dominos, 7) array.
    adjacent_doms : np.ndarray
        (num_dominos, num_dominos) relation matrix (will be modified in-place).

    Returns
    -------
    rows : list of np.ndarray
        Each element is a (2, n) array: row 0 = domino indices, row 1 = orientations.
    sums : np.ndarray
        Sum of dot values for each row.
    row_ends : np.ndarray
        (num_rows, 2) array with [first_end_value, last_end_value].
    """
    num_dominos = dominos.shape[0]
    rows = []

    for i in range(num_dominos):
        curr_dom = i
        left = i
        orient = -1
        next_orient = -1

        # Move left until we reach the beginning of the row
        while left >= 0:
            curr_dom = left
            orient = next_orient
            this_dom = dominos[left]

            if this_dom[4] == 1:  # Horizontal
                left, next_orient = left_dom(dominos, adjacent_doms, curr_dom, orient)
                if left >= 0:
                    curr_dom = left
            else:  # Vertical
                if orient != -1:
                    left, next_orient = left_dom(dominos, adjacent_doms, curr_dom, orient)
                else:
                    left, next_orient = left_dom(dominos, adjacent_doms, curr_dom, 2)
                    if left < 0:
                        left, next_orient = left_dom(dominos, adjacent_doms, curr_dom, 3)
                    if left < 0:
                        left, next_orient = left_dom(dominos, adjacent_doms, curr_dom, 4)

        # Move right until we reach the end of the row
        this_row_doms = [curr_dom]
        this_row_orients = []

        if orient != -1:
            this_row_orients.append(orient)
            right, next_orient = right_dom(dominos, adjacent_doms, curr_dom, orient)
        else:
            right, next_orient = right_dom(dominos, adjacent_doms, curr_dom, 1)
            this_row_orients.append(1)

            if right < 0:
                right, next_orient = right_dom(dominos, adjacent_doms, curr_dom, 2)
                this_row_orients[-1] = 2

            if right < 0:
                right, next_orient = right_dom(dominos, adjacent_doms, curr_dom, 3)
                this_row_orients[-1] = 3

            if right < 0:
                right, next_orient = right_dom(dominos, adjacent_doms, curr_dom, 4)
                this_row_orients[-1] = 4

        # Multi-domino row
        if right >= 0:
            while right >= 0:
                adjacent_doms[curr_dom, right] = 0
                adjacent_doms[right, curr_dom] = 0

                curr_dom = right
                orient = next_orient

                this_row_doms.append(curr_dom)
                this_row_orients.append(orient)

                right, next_orient = right_dom(dominos, adjacent_doms, curr_dom, orient)

            new_row = np.zeros((2, len(this_row_doms)), dtype=np.int32)
            for j in range(len(this_row_doms)):
                new_row[0, j] = this_row_doms[j]
                new_row[1, j] = this_row_orients[j]

            rows.append(new_row)

    num_rows = len(rows)
    sums = np.zeros(num_rows, dtype=np.int32)
    row_ends = np.zeros((num_rows, 2), dtype=np.int32)

    for i in range(num_rows):
        curr_row = rows[i]
        num_doms = curr_row.shape[1]
        row_sum = 0

        for j in range(num_doms):
            curr_d = dominos[curr_row[0, j]]
            curr_orient = curr_row[1, j]

            if curr_orient == 1:
                row_val = curr_d[5] + curr_d[6]
                start_val = curr_d[5]
                end_val = curr_d[6]
            elif curr_orient == 2:
                row_val = curr_d[5]
                start_val = curr_d[5]
                end_val = curr_d[5]
            elif curr_orient == 3:
                row_val = curr_d[6]
                start_val = curr_d[6]
                end_val = curr_d[6]
            else:  # orient == 4
                row_val = curr_d[5] + curr_d[6]
                start_val = curr_d[5]
                end_val = curr_d[5]

            row_sum += row_val

            if j == 0:
                row_ends[i, 0] = start_val
            if j == num_doms - 1:
                row_ends[i, 1] = end_val

        sums[i] = row_sum

    return rows, sums, row_ends
